exchanges/request/request.go

```diff
@@ -1,300 +1,152 @@
-		r.AuthLimit.SetRate(rate)
-		r.AuthLimit.SetDuration(duration)
-		return
-	}
-	r.UnauthLimit.SetRate(rate)
-	r.UnauthLimit.SetDuration(duration)
+	return err
 }
 
-// GetRateLimit gets the request Requester ratelimiter
-func (r *Requester) GetRateLimit(auth bool) *RateLimit {
-	if auth {
-		return r.AuthLimit
+// validateRequest validates the requester item fields
+func (i *Item) validateRequest(r *Requester) (*http.Request, error) {
+	if r == nil || r.Name == "" {
+		return nil, errors.New("not initialised, SetDefaults() called before making request?")
 	}
-	return r.UnauthLimit
-}
 
-// SetTimeoutRetryAttempts sets the amount of times the job will be retried
-// if it times out
-func (r *Requester) SetTimeoutRetryAttempts(n int) error {
-	if n < 0 {
-		return errors.New("routines.go error - timeout retry attempts cannot be less than zero")
+	if i == nil {
+		return nil, errors.New("request item cannot be nil")
 	}
-	r.timeoutRetryAttempts = n
-	return nil
-}
 
-// New returns a new Requester
-func New(name string, authLimit, unauthLimit *RateLimit, httpRequester *http.Client) *Requester {
-	return &Requester{
-		HTTPClient:           httpRequester,
-		UnauthLimit:          unauthLimit,
-		AuthLimit:            authLimit,
-		Name:                 name,
-		Jobs:                 make(chan Job, MaxRequestJobs),
-		timeoutRetryAttempts: TimeoutRetryAttempts,
-		timedLock:            timedmutex.NewTimedMutex(DefaultMutexLockTimeout),
+	if i.Path == "" {
+		return nil, errors.New("invalid path")
 	}
-}
-
-// IsValidMethod returns whether the supplied method is supported
-func IsValidMethod(method string) bool {
-	return common.StringDataCompareInsensitive(supportedMethods, method)
-}
 
-// IsValidCycle checks to see whether the current request cycle is valid or not
-func (r *Requester) IsValidCycle(auth bool) bool {
-	if auth {
-		if time.Since(r.Cycle) < r.AuthLimit.GetDuration() {
-			return true
-		}
-	} else {
-		if time.Since(r.Cycle) < r.UnauthLimit.GetDuration() {
-			return true
-		}
-	}
-
-	r.StartCycle()
-	return false
-}
-
-func (r *Requester) checkRequest(method, path string, body io.Reader, headers map[string]string) (*http.Request, error) {
-	req, err := http.NewRequest(method, path, body)
+	req, err := http.NewRequest(i.Method, i.Path, i.Body)
 	if err != nil {
 		return nil, err
 	}
 
-	for k, v := range headers {
+	for k, v := range i.Headers {
 		req.Header.Add(k, v)
 	}
 
-	if r.UserAgent != "" && req.Header.Get("User-Agent") == "" {
-		req.Header.Add("User-Agent", r.UserAgent)
+	if r.UserAgent != "" && req.Header.Get(userAgent) == "" {
+		req.Header.Add(userAgent, r.UserAgent)
 	}
 
 	return req, nil
 }
 
 // DoRequest performs a HTTP/HTTPS request with the supplied params
-func (r *Requester) DoRequest(req *http.Request, path string, body io.Reader, result interface{}, authRequest, verbose, httpDebug, httpRecord bool) error {
-	if verbose {
+func (r *Requester) DoRequest(req *http.Request, p *Item) error {
+	if p.Verbose {
```

Comment:
You'll need to nil check p else panic. Best do a `if p == nil { return nilErr }` style to minimise christmas trees

Reply:
I do the nill check in validateRequest method.

Reply:
have unexported that method so we only use SendPayload outside of this package

Reply:
I understand and thanks for making that change, but its still making an assumption that the only way that `doRequest` will be called is via `SendPayload`. If the only way its supposed to do that, then the only way to guarantee it that `p` is not nil is by putting the `doRequest` code inside `SendPayload`. Otherwise, please put in a nil check.